private/bufpkg/bufcheck/buflint/internal/buflintcheck/buflintcheck.go

```diff
@@ -963,3 +968,95 @@ func checkSyntaxSpecified(add addFunc, file protosource.File) error {
 	}
 	return nil
 }
+
+var (
+	// fieldConstraintsDesc provides a Descriptor for validate.FieldConstraints.
+	fieldConstraintsDesc = (*validate.FieldConstraints)(nil).ProtoReflect().Descriptor()
+
+	// fieldConstraintsOneOfDesc provides the OneOfDescriptor for the type union
+	// in FieldConstraints.
+	fieldConstraintsOneOfDesc = fieldConstraintsDesc.Oneofs().ByName("type")
+)
+
+// CheckValidateRulesTypesMatch is a check function.
+var CheckValidateRulesTypesMatch = newMessageCheckFunc(checkValidateRulesTypesMatch)
+
+func checkValidateRulesTypesMatch(add addFunc, message protosource.Message) error {
+	for _, field := range message.Fields() {
+		data, err := getDataByExtension(field, validate.E_Field)
+		if err != nil {
+			return err
+		}
+		var constraints validate.FieldConstraints
+		if err := proto.Unmarshal(data, &constraints); err != nil {
+			return fmt.Errorf("unmarshal error for field %q: %v", field.FullName(), err)
+		}
+		constraintsRefl := constraints.ProtoReflect()
+		setOneOf := constraintsRefl.WhichOneof(fieldConstraintsOneOfDesc)
+		if setOneOf == nil {
+			continue
+		}
+
+		got := field.Type()
+		expected := getTypeName(setOneOf.Name())
+		if got != expected {
+			add(field, field.Location(), nil, "constraint type mismatch on %q, expected %q but got %s", field.FullName(), expected, got)
+		}
+	}
+	return nil
+}
+
+func getDataByExtension(field protosource.OptionExtensionDescriptor, extensionType protoreflect.ExtensionType) ([]byte, error) {
+	extension, ok := field.OptionExtension(extensionType)
+	if !ok {
+		return nil, nil
+	}
+	fieldConstraints, ok := extension.(*dynamicpb.Message)
+	if !ok {
+		return nil, fmt.Errorf("unexpected extension type for field got %T", extension)
+	}
+	data, err := proto.Marshal(fieldConstraints)
+	if err != nil {
+		return nil, fmt.Errorf("marshal error for field: %v", err)
+	}
+	return data, nil
+}
+
+func getTypeName(in protoreflect.Name) descriptorpb.FieldDescriptorProto_Type {
```

Comment:
This isn't actually returning the type name (as the function implies), this is returning the `FieldDescriptorProto_Type` given the field name specific to protovalidate's library. When I first saw this I thought of it as something more generic, but it just happens that the field names match up with values in `FieldDescriptorProto_Type`. If we want a convenience function like this, it's something more like `getFieldDescriptorProtoTypeForConstFieldName`.

"Type name" typically refers to the message, enum, or group name for a given field: https://github.com/bufbuild/buf/blob/54ff2e4ebd2d0aa5b31010f22550d60649713cd4/private/bufpkg/bufcheck/bufbreaking/internal/bufbreakingcheck/bufbreakingcheck.go#L346

Reply:
I dont like the string compare here, its overly complex and unstable when it comes to data structures and wkts. I anticipate this function will not be here by the time this pr is ready for review